### PR TITLE
Implement IIListProvider on Distinct() and Union().

### PR DESCRIPTION
**Pull request by JonHanna**
Optimised `ToArray()` and `ToList()` on the results of these methods.

cc: @stephentoub @VSadov

### CODE REVIEW

File: src/System.Linq/src/System/Linq/Union.cs [hunk 1]
```diff
@@ -11,25 +12,130 @@ public static partial class Enumerable
...
+            }
+
+            public override bool MoveNext()
+            {
+                switch(state)
+                {
+                    case 1:
+                        _enumerator = _first.GetEnumerator();
+                        if (_enumerator.MoveNext())
+                            state = 2;
```
Comment (stephentoub) [inline src/System.Linq/src/System/Linq/Union.cs]: Please add braces here.

(In general, I don't like the style that's currently used through LINQ; it doesn't match what we've been trying to do in the rest of corefx.  Cleaning that up if @VSadov agrees should be a separate effort at some point.  In the meantime, though, in a case like this it just looks really strange to have an else with braces but the if without.)

**Comment by stephentoub** [comment]
LGTM

File: src/System.Linq/src/System/Linq/Union.cs (hunk 1 above)
Comment (JonHanna) [inline src/System.Linq/src/System/Linq/Union.cs]: Funnily enough while I do like brace-less two lines and hate the brace-less one-line used in this project, I've caught myself writing brace-less one-liners of late. This project is rubbing off on me.

Comment (VSadov) [inline src/System.Linq/src/System/Linq/Union.cs]: I also like the corefx style more. Except for the avoidance of "var". :-) 
We should run the formatter tool at some point on Linq and Expressions just for consistency.

Entered https://github.com/dotnet/corefx/issues/6150 for this.

**Comment by VSadov** [comment]
LGTM
but would wait for the braces change

**Comment by JonHanna** [comment]
Have fixed the braces.

A second commit here, not merged in yet so it can be dropped if you prefer, does a similar optimisation on chained unions to Stephen's for chained concatenations. While chained unions aren't as common, they aren't unheard of (the way to union 3+ sequences is to union and then union, after all) and this has the advantage of only having to create and fill a single hash set, rather than allocating and filling a new one for each call to `Union()`.

**Comment by stephentoub** [comment]
> A second commit here

Mind creating a separate PR?

**Comment by JonHanna** [comment]
Sure. Not being sure whether to include it in this or not is exactly why I didn't squash it in.

**Comment by JonHanna** [comment]
There, that's as the original was, bar the braces.

**Comment by JonHanna** [comment]
Test Innerloop Ubuntu Release Build and Test please

**Pull request merged**